File: 비트마스킹/실버/실버2/2961_도영이가 만든 맛있는 음식/2961_backtracking.java
import java.io.*;
import java.util.*;

public class Main {
    static int query;
    static int minNum = Integer.MAX_VALUE;
    static int[][] arr;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        query = Integer.parseInt(br.readLine().trim());
        arr = new int[query][2];
        for (int i = 0; i < query; i++) {
            String[] taste = br.readLine().split(" ");
            arr[i][0] = Integer.parseInt(taste[0]);
            arr[i][1] = Integer.parseInt(taste[1]);
        }

        findMinDifference(0, 1, 0, 0); //index, 신맛, 쓴맛, 재료사용횟수

        System.out.println(minNum);
    }

    public static void findMinDifference(int idx, int sour, int bitter, int used) {
        if (idx == query) {
            if (used > 0) {
                int difference = Math.abs(sour - bitter);
                minNum = Math.min(minNum, difference);
            }
            return;
        }

        findMinDifference(idx + 1, sour * arr[idx][0], bitter + arr[idx][1], used + 1);

        findMinDifference(idx + 1, sour, bitter, used);
    }
}

File: 비트마스킹/실버/실버2/2961_도영이가 만든 맛있는 음식/2961_backtracking.py
import sys
input = sys.stdin.read

# 전역 변수 선언
query = 0
min_num = float('inf')
arr = []

def find_min_difference(idx, sour, bitter, used):
    global min_num
    if idx == query:
        if used > 0:
            difference = abs(sour - bitter)
            min_num = min(min_num, difference)
        return

    # 재료를 사용한 경우
    find_min_difference(idx + 1, sour * arr[idx][0], bitter + arr[idx][1], used + 1)
    # 재료를 사용하지 않은 경우
    find_min_difference(idx + 1, sour, bitter, used)

def main():
    global query, arr
    data = input().strip().splitlines()
    query = int(data[0].strip())
    arr = [list(map(int, line.split())) for line in data[1:query + 1]]

    # 재귀 함수 호출
    find_min_difference(0, 1, 0, 0)

    # 결과 출력
    print(min_num)

if __name__ == "__main__":
    main()
